File: Plotter/src/GUI/raster/Raster.scala
package GUI.raster

import GUI.raster.path._

class Raster extends javax.swing.JPanel {
  case class Graph(val funk: mathParser.sprache.mathe.Real.Term, val variable: mathParser.sprache.mathe.Real.Variable) {
    def createTransformedShape(toDrawPlane: java.awt.geom.AffineTransform, config: modell.Viewport): java.awt.Shape = {
      val x_start = config.U.x
      val x_end = x_start + config.A.x
      val x_delta = config.A.x / 1000d;

      val path = new java.awt.geom.Path2D.Double();
      variable(x_start)
      path.moveTo(x_start, funk())
      for (x <- Range.Double(x_start, x_end, x_delta)) {
        variable(x);
        val y = funk();
        if (java.lang.Double.isNaN(y) || java.lang.Double.isInfinite(y)) {
          path.moveTo(x, y);
        } else {
          path.lineTo(x, y);
        }
      }
      return path.createTransformedShape(toDrawPlane);
    }
  }

  var view = new modell.Viewport(new modell.Point(-10, -10), 20, 20)
  var graphs = List[Graph]()

  super.setPreferredSize(new java.awt.Dimension(750, 750))
  super.addKeyListener(new java.awt.event.KeyAdapter {
    override def keyPressed(arg: java.awt.event.KeyEvent) = {
      arg.getKeyCode() match {
        case 38 => view = view.up()
        case 39 => view = view.right()
        case 40 => view = view.down()
        case 37 => view = view.left()
        case 107 => view = view.zoomIn()
        case 108 => view = view.zoomOut()
      }
      repaint()
    }
  })
  super.addMouseListener(new java.awt.event.MouseAdapter {
    override def mousePressed(e: java.awt.event.MouseEvent) = {
      val x = e.getX().asInstanceOf[Double] / getWidth()
      val y = e.getY().asInstanceOf[Double] / getHeight()
      view = view.focus(x, y)
      repaint()
    }
  })
  super.addMouseWheelListener(new java.awt.event.MouseAdapter() {
    override def mouseWheelMoved(e: java.awt.event.MouseWheelEvent) = {
      val x = e.getX().asInstanceOf[Double] / getWidth()
      val y = e.getY().asInstanceOf[Double] / getHeight()
      view = view.zoom(x, y, e.getWheelRotation());
      repaint();
    }
  });
  super.setFocusable(true);

  override def paint(g: java.awt.Graphics): Unit = {
    if (!g.isInstanceOf[java.awt.Graphics2D])
      return
    val graphics = g.asInstanceOf[java.awt.Graphics2D]
    graphics.setRenderingHint(java.awt.RenderingHints.KEY_ANTIALIASING, java.awt.RenderingHints.VALUE_ANTIALIAS_ON)

    graphics.setBackground(java.awt.Color.white)
    graphics.clearRect(0, 0, getWidth(), getHeight())

    val trans = new modell.Transform(view, new modell.Dimensions(getWidth(), getHeight()))

    val Atrans = trans.asTransform()
    try {
      Atrans.invert();
    } catch {
      case e: java.awt.geom.NoninvertibleTransformException => return
    }

    // graphics.setColor(Color.lightGray);
    // graphics.draw(new Background().createTransformedShape(Atrans, view));
    graphics.setColor(java.awt.Color.black);
    graphics.draw(new Axes().createTransformedShape(Atrans, view));

    graphics.setColor(java.awt.Color.red);
    for (graph <- graphs) {
      graphics.draw(graph.createTransformedShape(Atrans, view));
    }
  }

  def addFunction(funk: mathParser.sprache.mathe.Real.Term, variable: mathParser.sprache.mathe.Real.Variable) = {
    graphs = new Graph(funk, variable) :: graphs
    repaint()
  }

  def clear() = {
    graphs = Nil
    repaint()
  }

    import mathParser.sprache.mathe._
    import mathParser.sprache.mathe.Real._
  
    def parseAndDrawFunction(term: String, variable: String):Unit = {
      val p = new mathParser.parser.Parser(Real)
      val x = p.registerVariable(variable)
      p(term) match {
        case Some(t) => {
          addFunction(t.asInstanceOf[Term], x.asInstanceOf[Variable])
        }
        case None => 
      }
    }

}

File: Plotter/src/GUI/raster/Raster.java
package GUI.raster;
//package GUI.raster;
//
//import java.awt.Color;
//import java.awt.Dimension;
//import java.awt.Graphics;
//import java.awt.Graphics2D;
//import java.awt.RenderingHints;
//import java.awt.event.KeyAdapter;
//import java.awt.event.KeyEvent;
//import java.awt.event.MouseAdapter;
//import java.awt.event.MouseEvent;
//import java.awt.event.MouseWheelEvent;
//import java.awt.geom.AffineTransform;
//import java.awt.geom.NoninvertibleTransformException;
//import java.util.ArrayList;
//
//import javax.swing.JPanel;
//
//import term.IFunction;
//import term.variable.Variable;
//import GUI.raster.path.Axes;
//import GUI.raster.path.Background;
//import GUI.raster.path.Graph;
//
//@SuppressWarnings("serial")
//public class Raster extends JPanel {
//	// Konfiguration
//	private modell.Viewport view = new modell.Viewport(new modell.Point(-10, -10), 20., 20.);
//	private static Color colorBg = new Color(255, 255, 255);
//	private static int startWidth = 750, startHeight = 750;
//	private final ArrayList<Graph> graphs = new ArrayList<Graph>();
//
//	public Raster() {
//		setPreferredSize(new Dimension(startWidth, startHeight));
//		addKeyListener(new KeyAdapter() {
//			@Override
//			public void keyPressed(KeyEvent arg0) {
//				switch (arg0.getKeyCode()) {
//				case 38:
//					view = view.up();
//					break;
//				case 39:
//					view = view.right();
//					break;
//				case 40:
//					view = view.down();
//					break;
//				case 37:
//					view = view.left();
//					break;
//				case 107:
//					view = view.zoomIn();
//					break;
//				case 109:
//					view = view.zoomOut();
//					break;
//				default:
//					return;
//				}
//				repaint();
//			}
//		});
//
//		addMouseListener(new MouseAdapter() {
//			public void mousePressed(MouseEvent e) {
//				double x = e.getX() / (double) getWidth();
//				double y = e.getY() / (double) getHeight();
//				view = view.focus(x, y);
//				repaint();
//			}
//		});
//
//		addMouseWheelListener(new MouseAdapter() {
//			public void mouseWheelMoved(MouseWheelEvent e) {
//				double x = e.getX() / (double) getWidth();
//				double y = e.getY() / (double) getHeight();
//				view = view.zoom(x, y, e.getWheelRotation());
//				repaint();
//			}
//		});
//
//		setFocusable(true);
//	}
//
//	@Override
//	public void paint(Graphics g_) {
//		Graphics2D graphics = (Graphics2D) g_;
//		graphics.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
//
//		graphics.setBackground(colorBg);
//		graphics.clearRect(0, 0, getWidth(), getHeight());
//
//		modell.Transform trans = new modell.Transform(view, new modell.Dimensions(getWidth(), getHeight()));
//
//		AffineTransform Atrans = trans.asTransform();
//		try {
//			Atrans.invert();
//		} catch (NoninvertibleTransformException e) {
//			// TODO Auto-generated catch block
//			e.printStackTrace();
//		}
//
//		// graphics.setColor(Color.lightGray);
//		// graphics.draw(new Background().createTransformedShape(Atrans, view));
//		graphics.setColor(Color.black);
//		graphics.draw(new Axes().createTransformedShape(Atrans, view));
//
//		graphics.setColor(Color.red);
//		for (Graph g : graphs) {
//			graphics.draw(g.createTransformedShape(Atrans, view));
//		}
//	}
//
//	public void addFunction(IFunction function, Variable variable, Color color) {
//		graphs.add(new Graph(function, variable));
//		repaint();
//	}
//
//	public void clear() {
//		graphs.clear();
//		repaint();
//	}
//}
